### ISSUE TITLE
[DTDLv3] Provide native extensions for invariants on primitive data types

### ISSUE THREAD

**Issue by thern743** [Azure/opendigitaltwins-dtdl#156]
IoT Central has built-in schema extensions for primitive types. These allow for metadata for invariants such as default values, minimum values, maximum values, etc.

How does Microsoft propose we adopt this into the wider ecosystem? Implementing these ourselves seems counter-productive and trying to use the new Annotations seems unnecessarily verbose.

Can we propose a native extension that include these out of the box in V3?

**Comment by thern743** [comment]
Yes, I think that's exactly what I would expect. I was confused because the DTDL v3 documentation specifically mentions *partner* extension and calls out IoT Central specifically.

**Comment by jrdouceur** [comment]
Where are you seeing a reference to partner extensions or IoT Central in the DTDL v3 docs?  The [main v3 doc](https://github.com/Azure/opendigitaltwins-dtdl/blob/master/DTDL/v3/DTDL.v3.md) does not mention either of these, and the [extensions doc](https://github.com/Azure/opendigitaltwins-dtdl/blob/master/DTDL/v3/DTDL.Extensions.md) shows the Central partner extension associated with DTDL v2 and the feature extensions associated with DTDL v3.

**Comment by thern743** [comment]
Ahh, I see. It wasn't clear to me that the "partner extensions" from IoT Central were only available in V2 (I see it in the table). The language at the top made it seem like "partner" extensions were a feature of DTDLv3:

> DTDL language extensions -- PREVIEW...There are two categories of language extensions: Partner extensions add types, properties, and elements that are relevant to a specific cloud service. Feature extensions add non-core features to the DTDL language in a selectable manner.

**Comment by jrdouceur** [comment]
This was not intended to be a DTDLv3-specific document. It is a general DTDL doc describing the concept of language extensions.  There is no indication in the doc title or header of DTDL v3.  However, you might have been confused by the placement of this doc under the "v3" folder.  There was some disagreement about the appropriate location of some of the doc files, and perhaps this one landed in a sub-optimal location that led to an incorrect inference.

**Comment by thern743** [comment]
Thanks @jrdouceur, makes sense. While we're on the topic, do you foresee special extensions for the concepts of _Cloud Properties_ as implemented in IoT Central? Or do we see this as a specific type of annotation?

**Comment by jrdouceur** [comment]
This is not something I've been involved in discussing, so I don't know whether there are any plans either way.  My knee-jerk opinion is that the impact of co-typing a Property with Cloud is comparable to that of co-typing with [Streaming](https://github.com/Azure/opendigitaltwins-dtdl/blob/master/DTDL/v3/DTDL.streaming.v1.md), so this would argue for a distinct adjunct type rather than an annotation.  But I do not know of any plans regarding this as a feature in any form outside of IoT Central.
